drivers/clock_control/clock_control_nrfs_audiopll.c

```diff
@@ -0,0 +1,275 @@
+ *
+ * Simplified linear approximation:
+ *
+ *   frequency = 10666667 + (((13333293 - 10666667) / 65535) * freq_fraction)
+ *   frequency = 10666667 + ((2666626 / 65535) * freq_fraction)
+ *   frequency = ((10666667 * 65535) + (2666626 * freq_fraction)) / 65535
+ *   frequency = (699040021845 + (2666626 * freq_fraction)) / 65535
+ *
+ * Isolate freq_fraction:
+ *
+ *   frequency = (699040021845 + (2666626 * freq_fraction)) / 65535
+ *   frequency * 65535 = 699040021845 + (2666626 * freq_fractio)
+ *   (frequency * 65535) - 699040021845 = 2666626 * freq_fractio
+ *   freq_fractio = ((frequency * 65535) - 699040021845) / 2666626
+ */
+static uint16_t shim_frequency_to_freq_fraction(uint32_t frequency)
+{
+	uint64_t freq_fraction;
+
+	freq_fraction = frequency;
+	freq_fraction *= 65535;
+	freq_fraction -= 699040021845;
+	freq_fraction /= 2666626;
+
+	return (uint16_t)freq_fraction;
+}
+
+static uint32_t shim_freq_fraction_to_frequency(uint16_t freq_fraction)
+{
+	uint64_t frequency;
+
+	frequency = 2666626;
+	frequency *= freq_fraction;
+	frequency += 699040021845;
+	frequency /= 65535;
+
+	return (uint32_t)frequency;
+}
+
+static bool shim_frequency_valid(uint32_t frequency)
+{
+	return frequency >= SHIM_MIN_FREQUENCY && frequency <= SHIM_MAX_FREQUENCY;
+}
+
+static int shim_clock_control_api_on(const struct device *dev,
+				     clock_control_subsys_t sub_system)
+{
+	ARG_UNUSED(sub_system);
+
+	return shim_nrfs_request_enable(dev);
+}
+
+static int shim_clock_control_api_off(const struct device *dev,
+				      clock_control_subsys_t sub_system)
+{
+	ARG_UNUSED(sub_system);
+
+	return shim_nrfs_request_disable(dev);
+}
+
+static int shim_clock_control_api_set_rate(const struct device *dev,
+					   clock_control_subsys_t sub_system,
+					   clock_control_subsys_rate_t rate)
+{
+	struct nrf_clock_control_audiopll_rate *audiopll_rate = rate;
+	uint16_t freq_fraction;
+
+	ARG_UNUSED(sub_system);
+
+	if (!shim_frequency_valid(audiopll_rate->frequency)) {
+		return -EINVAL;
+	}
+
+	freq_fraction = shim_frequency_to_freq_fraction(audiopll_rate->frequency);
+	return shim_nrfs_request_freq(dev, freq_fraction);
+}
+
+static int shim_clock_control_api_get_rate(const struct device *dev,
+					   clock_control_subsys_t sub_system,
+					   uint32_t *rate)
+{
+	struct shim_data *dev_data = dev->data;
+
+	ARG_UNUSED(sub_system);
+
+	*rate = shim_freq_fraction_to_frequency(dev_data->freq_fraction);
+	return 0;
+}
+
+static DEVICE_API(clock_control, shim_clock_control_api) = {
+	.on = shim_clock_control_api_on,
+	.off = shim_clock_control_api_off,
+	.set_rate = shim_clock_control_api_set_rate,
+	.get_rate = shim_clock_control_api_get_rate,
+};
+
+static int shim_init(const struct device *dev)
+{
+	struct shim_data *dev_data = dev->data;
+	const struct shim_config *dev_config = dev->data;
+	nrfs_err_t err;
+	int ret;
+
+	LOG_DBG("waiting for nrfs backend connected");
+	err = nrfs_backend_wait_for_connection(K_FOREVER);
```

Comment:
Maybe it would be better not to wait forever? Usually this should take less than a 1s to connect (maybe put here 5s timeout) If it doesn't connect in 5s then there is something wrong, but we do not hang here. What do You think?

Reply:
If NRFS does not connect, we should not boot, regardless of the wait. K_FOREVER is the simplest way to achieve this :)

Reply:
@ahasztag Could You check, I remember that there is some work to change this?

Reply:
Ok this will be discussed later with Hubert, for now this is ok.